chips/ibex/src/chip.rs

```diff
@@ -0,0 +1,232 @@
+//! High-level setup and interrupt mapping for the chip.
+
+use kernel;
+use kernel::debug;
+use rv32i;
+use rv32i::csr;
+
+use crate::plic;
+use crate::timer;
+
+pub struct Ibex {
+    userspace_kernel_boundary: rv32i::syscall::SysCall,
+}
+
+impl Ibex {
+    pub unsafe fn new() -> Ibex {
+        Ibex {
+            userspace_kernel_boundary: rv32i::syscall::SysCall::new(),
+        }
+    }
+
+    pub unsafe fn enable_plic_interrupts(&self) {
+        plic::disable_all();
+        plic::clear_all_pending();
+        plic::enable_all();
+    }
+}
+
+impl kernel::Chip for Ibex {
+    type MPU = ();
+    type UserspaceKernelBoundary = rv32i::syscall::SysCall;
+    type SysTick = ();
+
+    fn mpu(&self) -> &Self::MPU {
+        &()
+    }
+
+    fn systick(&self) -> &Self::SysTick {
+        &()
+    }
+
+    fn userspace_kernel_boundary(&self) -> &rv32i::syscall::SysCall {
+        &self.userspace_kernel_boundary
+    }
+
+    fn service_pending_interrupts(&self) {
+        unsafe {
+            while let Some(interrupt) = plic::next_pending() {
+                match interrupt {
+                    _ => debug!("Pidx {}", interrupt),
+                }
+
+                // Mark that we are done with this interrupt and the hardware
+                // can clear it.
+                plic::complete(interrupt);
+            }
+        }
+    }
+
+    fn has_pending_interrupts(&self) -> bool {
+        unsafe { plic::has_pending() }
+    }
+
+    fn sleep(&self) {
+        unsafe {
+            rv32i::support::wfi();
+        }
+    }
+
+    unsafe fn atomic<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce() -> R,
+    {
+        rv32i::support::atomic(f)
+    }
+}
+
+pub unsafe fn handle_trap() {
+    let cause = rv32i::csr::CSR.mcause.extract();
+
+    // if most sig bit is set, is interrupt
+    // strip off the msb
+    match rv32i::csr::mcause::McauseHelpers::cause(&cause) {
```

Comment:
I think this structure is fine for now, but at some point we should look at the generated assembly. This is a case where I'm personally unsure how the Rust will map to assembly code and I worry it might be inefficient. Since this is the core trap handler that matters.

Reply:
Good point